libcontainer/utils/utils_unix.go

```diff
@@ -23,9 +26,44 @@ func EnsureProcHandle(fh *os.File) error {
 	return nil
 }
 
+var (
+	haveCloseRangeCloexec     bool
+	haveCloseRangeCloexecOnce sync.Once
+)
+
+func tryCloseRangeCloexec(minFd int) error {
+	haveCloseRangeCloexecOnce.Do(func() {
+		// Make sure we're not closing a random file descriptor.
+		tmpFd, err := unix.FcntlInt(0, unix.F_DUPFD_CLOEXEC, 0)
+		if err != nil {
+			return
+		}
+		defer unix.Close(tmpFd)
+
+		err = unix.CloseRange(uint(tmpFd), uint(tmpFd), unix.CLOSE_RANGE_CLOEXEC)
+		// Any error means we cannot use close_range(CLOSE_RANGE_CLOEXEC).
+		// -ENOSYS and -EINVAL ultimately mean we don't have support, but any
+		// other potential error would imply that even the most basic close
+		// operation wouldn't work.
+		haveCloseRangeCloexec = err == nil
+	})
+
+	var err error
+	if haveCloseRangeCloexec {
+		err = unix.CloseRange(uint(minFd), math.MaxUint, unix.CLOSE_RANGE_CLOEXEC)
+	} else {
+		err = unix.ENOSYS
+	}
+	return &os.SyscallError{Syscall: "close_range", Err: err}
+}
+
 // CloseExecFrom applies O_CLOEXEC to all file descriptors currently open for
 // the process (except for those below the given fd value).
 func CloseExecFrom(minFd int) error {
+	if err := tryCloseRangeCloexec(minFd); !errors.Is(err, unix.ENOSYS) {
```

Comment:
I think we should check `haveCloseRangeCloexec` directly here, otherwise the code will fail on some ugly setups (seccomp returning a wrong error code, etc.)

While at it, we can wrap the error right here.

So, something like

```go
func haveCloseRangeCloexec() bool {
	haveCloseRangeCloexecOnce.Do(func() {
		// Make sure we're not closing a random file descriptor.
		tmpFd, err := unix.FcntlInt(0, unix.F_DUPFD_CLOEXEC, 0)
		if err != nil {
			return
		}
		defer unix.Close(tmpFd)

		err = unix.CloseRange(uint(tmpFd), uint(tmpFd), unix.CLOSE_RANGE_CLOEXEC)
		// Any error means we cannot use close_range(CLOSE_RANGE_CLOEXEC).
		// -ENOSYS and -EINVAL ultimately mean we don't have support, but any
		// other potential error would imply that even the most basic close
		// operation wouldn't work.
		haveCloseRangeCloexec = err == nil
	})
	
	return haveCloseRangeCloexec
}

// CloseExecFrom applies O_CLOEXEC to all file descriptors currently open for
// the process (except for those below the given fd value).
func CloseExecFrom(minFd int) error {
	if haveCloseRangeCloexec() {
		err := unix.CloseRange(uint(minFd), math.MaxUint, unix.CLOSE_RANGE_CLOEXEC)
		if err != nil {
			return &os.SyscallError{Syscall: "close_range", Err: err}
		}
		return nil
	}
	// Fallback to old implementation.

```

Reply:
I think even in ugly seccomp scenarios, either way should work, but I guess `have...()` is closer to how we do this elsewhere, so I'll switch it over.